pkg/agent/packetcapture/capture/bpf.go

```diff
@@ -92,23 +172,70 @@ func loadIPv4HeaderOffset(skipTrue uint8) []bpf.Instruction {
 	}
 }
 
-func compareProtocolIP4(skipTrue, skipFalse uint8) bpf.Instruction {
-	return bpf.JumpIf{Cond: bpf.JumpEqual, Val: etherTypeIPv4, SkipTrue: skipTrue, SkipFalse: skipFalse}
+func compareProtocolIP(etherType uint32, skipTrue, skipFalse uint8) bpf.Instruction {
+	return bpf.JumpIf{Cond: bpf.JumpEqual, Val: etherType, SkipTrue: skipTrue, SkipFalse: skipFalse}
 }
 
 func compareProtocol(protocol uint32, skipTrue, skipFalse uint8) bpf.Instruction {
 	return bpf.JumpIf{Cond: bpf.JumpEqual, Val: protocol, SkipTrue: skipTrue, SkipFalse: skipFalse}
 }
 
-func calculateSkipFalse(srcIP, dstIP net.IP, transport *transportFilters) uint8 {
+// getAddressChunk abstracts the process of extracting a 4-byte chunk from an IP address,
+// handling the structural differences between IPv4 (one chunk) and IPv6 (four chunks).
+func (h *ipFamilyHandler) getAddressChunk(ip net.IP, chunkIndex int) uint32 {
+	if h.etherType == etherTypeIPv4 {
+		return binary.BigEndian.Uint32(ip[len(ip)-4:])
+	}
+	return binary.BigEndian.Uint32(ip[chunkIndex*4 : (chunkIndex+1)*4])
+}
+
+// calculateSkipOffset determines the correct 'SkipFalse' jump offset for an IP address chunk
+// comparison. When checking bidirectional traffic ('Both' direction), a failed check for the first
+// direction should not jump to the end (drop), but rather to the start of the check for the other
+// direction. jumpToReturnTraffic: If true, calculate the offset to jump to the return traffic block.
+// chunkIndex: The current 4-byte chunk index of the IP being checked (0-3 for IPv6).
+func (h *ipFamilyHandler) calculateSkipOffset(chunkIndex int, skipFalse, skipToEnd uint8, jumpToReturnTraffic bool) uint8 {
+	if jumpToReturnTraffic {
+		if h.etherType == etherTypeIPv4 {
+			return skipFalse
+		}
+		// calculate the relative jump offsets (SkipFalse) that decrease by 2 per chunk
+		// for the srcIP and dstIP cases.
+		return skipFalse - uint8((chunkIndex+1)*2)
+	}
+	return skipToEnd
+}
+
+func (h *ipFamilyHandler) countAddrInstForSkipFalse(srcIP, dstIP net.IP) uint8 {
```

Comment:
nit: would not abbreviate `Inst` here, that's inconsistent with the rest of the code